modules/java/jar/CMakeLists.txt

```diff
@@ -34,19 +34,41 @@ if(OPENCV_JAVADOC_LINK_URL)
   set(CMAKE_CONFIG_OPENCV_JAVADOC_LINK "link=\"${OPENCV_JAVADOC_LINK_URL}\"")
 endif()
 
-configure_file("${CMAKE_CURRENT_SOURCE_DIR}/build.xml.in" "${OPENCV_JAVA_DIR}/build.xml" @ONLY)
-list(APPEND depends "${OPENCV_JAVA_DIR}/build.xml")
-
-ocv_cmake_byproducts(__byproducts BYPRODUCTS "${OPENCV_JAR_FILE}")
-add_custom_command(OUTPUT "${OPENCV_DEPHELPER}/${the_module}_jar"
-    ${__byproducts}  # required for add_custom_target() by ninja
-    COMMAND ${ANT_EXECUTABLE} -noinput -k jar
-    COMMAND ${CMAKE_COMMAND} -E touch "${OPENCV_DEPHELPER}/${the_module}_jar"
-    WORKING_DIRECTORY "${OPENCV_JAVA_DIR}"
-    DEPENDS ${depends}
-    COMMENT "Generating ${JAR_NAME}"
-)
-add_custom_target(${the_module}_jar DEPENDS "${OPENCV_DEPHELPER}/${the_module}_jar")
+if (ANT_EXECUTABLE)
+  configure_file("${CMAKE_CURRENT_SOURCE_DIR}/build.xml.in" "${OPENCV_JAVA_DIR}/build.xml" @ONLY)
+  list(APPEND depends "${OPENCV_JAVA_DIR}/build.xml")
+
+  ocv_cmake_byproducts(__byproducts BYPRODUCTS "${OPENCV_JAR_FILE}")
+  add_custom_command(OUTPUT "${OPENCV_DEPHELPER}/${the_module}_jar"
+      ${__byproducts}  # required for add_custom_target() by ninja
+      COMMAND ${ANT_EXECUTABLE} -noinput -k jar
+      COMMAND ${CMAKE_COMMAND} -E touch "${OPENCV_DEPHELPER}/${the_module}_jar"
+      WORKING_DIRECTORY "${OPENCV_JAVA_DIR}"
+      DEPENDS ${depends}
+      COMMENT "Generating ${JAR_NAME}"
+  )
+  add_custom_target(${the_module}_jar DEPENDS "${OPENCV_DEPHELPER}/${the_module}_jar")
+else()
+  find_package(Java REQUIRED)
+  include(UseJava)
+
+  add_custom_target(${the_module}_jar_sources
+                    COMMAND ${CMAKE_COMMAND}
+                    -D OPENCV_JAVA_DIR="${OPENCV_JAVA_DIR}"
+                    -D OUTPUT="${OPENCV_JAVA_DIR}/java_sources"
+                    -P "${CMAKE_CURRENT_SOURCE_DIR}/list_java_sources.cmake")
+
+  get_filename_component(OPENCV_JAR_DIR ${OPENCV_JAR_FILE} DIRECTORY)
+  get_filename_component(JAR_NAME_WE ${JAR_NAME} NAME_WE)
+
+  add_jar(${the_module}_jar
+          SOURCES "@${OPENCV_JAVA_DIR}/java_sources"
```

Comment:
requires CMake 3.4, so probably need to add `cmake_minimum_required`?

Reply:
OpenCV 4.x requires CMake 3.5.1 (base version for Ubuntu 16.04)